docs/specs/lcb-signals-and-events.md

```diff
@@ -0,0 +1,240 @@
+
+A pre-event handler is declared using a `pre event handler` declaration:
+
+    pre event handler for <name:id> is <handler:constexpr>
+
+By default, all pre-event handlers declared by the list of interested controls are run.  There is no need to "pass" the event onwards [1].
+
+A control can use the pre-event handler to **capture** an event:
+
+    capture event
+
+The `capture event` statement makes the capturing control becomes the "target" for the event; all its children are removed from the list of interested controls [2,3].
+
+**Notes:**
+
+1. As per the `mouseEnter`/`mouseLeave`, implicit passing reflects the expectation that pre- and post-event handlers are usually be used to implement behaviour that complements rather than conflicts with that of other interested controls.
+
+2. This allows composed widgets to "hide" events from their children.
+
+3. Since `capture event` causes the capturing control to become the target, the capturing control can't do anything that requires knowledge of the "original" target in its on-event handler unless it saves `the event target` in the pre-event handler before `capture event`.  On the other hand, there's not much point in having separate pre-event and on-event handlers when capturing because they're guaranteed to be run immediately one after another.
+
+### Dispatch phase
+
+During the dispatch phase, **on-event** handlers are run.  They are run from "inside-out", starting from the target control.  A parent widget's on-event handler runs after its child widgets' on-event handlers.
+
+An on-event handler is declared using an `event handler` declaration [1]:
+
+    [on] event handler for <name:id> is <handler:constexpr>
+
+By default, if an on-event handler is found for an event, then it **blocks** further on-event handler execution for the event [2].
+
+An on-event handler can use the `pass event` statement to prevent the default-blocking behaviour.
+
+The `pass event` statement does not return.
+
+If an on-event handler yields, it is treated as having blocked further on-event handler execution.  It is an error to `pass event` after yielding [3].
+
+**Notes:**
+
+1. Since most widgets will only need to handle events during the dispatch phase, on-event handlers may be declared without a qualifier.
+
+2. This is for conceptual compatibility with LCS.
+
+3. Since yielding detaches an event handler from event processing, there is subsequently no "next handler" for the on-event handler to pass the event on to.
+
+### Leave phase
+
+During the leave phase, **post-event** handlers are run.  Like on-event handlers, they are run from "inside-out", starting from the target control.
+
+All interested controls have the opportunity to run a post-event handler, even if an on-event handler blocked the event during the dispatch phase [1].
+
+A post-event handler is declared using a `post event handler` declaration:
+
+    post event handler for <name:id> is <handler:constexpr>
+
```

Comment:
It wasn't lost!

Unfortunately, and as you quite rightly pointed out in a previous comment, the `before` and `after` handlers in LCS work have quite a different semantic that relates to behaviours and not really specifically to events.  @runrevmark and I had a long discussion about whether it was possible to rejig the way they work so that they can be used to do "pre" and "post" event handling.  Unfortunately, we couldn't find a way to make that sane, so for the time being I think the plan is to make "pre" and "post" event handling available only in LCB.

This also gives the opportunity to check that the event model makes sense, since we are reserving the right to make backwards-incompatible changes to LCB in the future (but we don't have that luxury for LCS).